src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2702,19 +2702,52 @@ private int LearnFromNullTest(BoundExpression expression, ref LocalState state)
             // We should not blindly strip conversions here. Tracked by https://github.com/dotnet/roslyn/issues/36164
             var expressionWithoutConversion = RemoveConversion(expression, includeExplicitConversions: true).expression;
             var slot = MakeSlot(expressionWithoutConversion);
-            return LearnFromNullTest(slot, expressionWithoutConversion.Type, ref state);
+
+            // since we know for sure the slot is null (we just tested it), we know that dependent slots are not
+            // reachable and therefore can be treated as not null.  However, we have not computed the proper
+            // (inferred) type for the expression, so we cannot compute the correct symbols for the member slots here
+            // (using the incorrect symbols would result in computing an incorrect default state for them).
+            // Therefore we do not mark dependent slots not null.  See https://github.com/dotnet/roslyn/issues/39624
+            return LearnFromNullTest(slot, expressionWithoutConversion.Type, ref state, markDependentSlotsNotNull: false);
```

Comment:
>markDependentSlotsNotNull: false [](start = 88, length = 32)

Consider either making the parameter always explicit, or not passing the default value.

Reply:
Alternatively, since there's only one call that actually passes true, consider not adding this flag to `LearnFromNullTest` and just calling `MarkDependentSlotsNotNull` there directly.

---
In reply to: [357310030](https://github.com/dotnet/roslyn/pull/39625#discussion_r357310030) [](ancestors = 357310030)